File: examples/dekita/.claude/scripts/ci_monitor/worktree.py
"""Worktree management for ci-monitor.

This module handles git worktree detection and cleanup operations.
Extracted from ci-monitor.py as part of Issue #1754 refactoring.
"""

from __future__ import annotations

import os
import re
import subprocess
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Callable


def get_worktree_info() -> tuple[str | None, str | None]:
    """Get worktree information if current directory is inside a worktree.

    Returns:
        Tuple of (main_repo_path, worktree_path) if in worktree, (None, None) otherwise.
    """
    try:
        # Get current working directory
        cwd = os.getcwd()

        # Get git worktree list
        result = subprocess.run(
            ["git", "worktree", "list", "--porcelain"],
            capture_output=True,
            text=True,
            timeout=10,
        )
        if result.returncode != 0:
            return None, None

        # Parse worktree list
        # Format: worktree /path\nHEAD abc123\nbranch refs/heads/xxx\n\n
        worktrees: list[str] = []
        main_repo: str | None = None
        for line in result.stdout.strip().split("\n"):
            if line.startswith("worktree "):
                path = line[9:]  # Remove "worktree " prefix
                if main_repo is None:
                    main_repo = path  # First entry is main repo
                else:
                    worktrees.append(path)

        if main_repo is None:
            return None, None

        # Check if cwd is inside a worktree (not main repo)
        cwd_real = os.path.realpath(cwd)
        for wt_path in worktrees:
            wt_real = os.path.realpath(wt_path)
            if cwd_real == wt_real or cwd_real.startswith(wt_real + os.sep):
                return main_repo, wt_path

        return None, None
    except Exception:  # noqa: BLE001 - Catch all for subprocess errors
        return None, None


def _is_exact_worktree_match(branch: str, wt_name: str) -> bool:
    """Check if branch name contains exact worktree name as a segment.

    Examples:
        - "fix/issue-1366-cleanup" matches "issue-1366" -> True
        - "fix/issue-13669" matches "issue-1366" -> False
        - "feature-1234" matches "123" -> False

    Args:
        branch: Branch name to check.
        wt_name: Worktree name to match.

    Returns:
        True if the branch contains the exact worktree name as a segment.
    """
    # Check for exact segment match (surrounded by non-alphanumeric or at boundaries)
    # Escape special regex characters in wt_name
    escaped = re.escape(wt_name)
    # Match only if wt_name is a complete segment (not part of a larger number/word)
    pattern = rf"(^|[^a-zA-Z0-9]){escaped}([^a-zA-Z0-9]|$)"
    return bool(re.search(pattern, branch))


def cleanup_worktree_after_merge(
    wt_path: str,
    main_repo: str,
    *,
    json_mode: bool = False,
    log_fn: Callable[[str, bool, dict[str, Any] | None], None] | None = None,
) -> bool:
    """Cleanup worktree after successful merge.

    Issue #1366: Prevent worktree from remaining after cleanup phase is skipped.

    Note: Issue #1478 - This function is no longer called from --merge command
    because subprocess deletion makes parent process's cwd invalid. However, it
    is kept for potential use by hooks (e.g., worktree-auto-cleanup.py) that run
    in the correct process context.

    Args:
        wt_path: Path to the worktree to remove.
        main_repo: Path to the main repository.
        json_mode: If True, log() outputs to stderr instead of stdout (for JSON output).
        log_fn: Optional logging function. Signature: (message, json_mode, data) -> None.

    Returns:
        True if cleanup succeeded, False otherwise.
    """

    def _log(message: str) -> None:
        """Internal log wrapper."""
        if log_fn:
            log_fn(message, json_mode, None)
        else:
            print(message)

    def _cleanup_branch(wt_name: str) -> None:
        """Delete local branch matching the worktree name (best-effort)."""
        try:
            # Get branches with exact issue number match (e.g., issue-1366)
            # Using --list with pattern, then filter for exact match
            branch_result = subprocess.run(
                ["git", "branch", "--list"],
                capture_output=True,
                text=True,
                timeout=10,
            )
            if branch_result.returncode == 0 and branch_result.stdout.strip():
                for line in branch_result.stdout.strip().split("\n"):
                    branch = line.strip().lstrip("* ")
                    # Exact match: branch must contain the full wt_name as a segment
                    # e.g., "fix/issue-1366-xxx" matches "issue-1366"
                    # but "fix/issue-13669" does NOT match "issue-1366"
                    if branch and _is_exact_worktree_match(branch, wt_name):
                        # -d is safe: only deletes fully merged branches
                        del_result = subprocess.run(
                            ["git", "branch", "-d", branch],
                            capture_output=True,
                            text=True,
                            timeout=10,
                        )
                        if del_result.returncode == 0:
                            _log(f"🗑️ ブランチ {branch} を削除しました")
        except Exception as e:  # noqa: BLE001 - Best-effort cleanup
            # Branch cleanup is best-effort - log error but don't fail
            _log(f"⚠️ ブランチ削除中にエラー: {e}")

    try:
        # If cwd is inside worktree, move to main repo first
        cwd = os.getcwd()
        cwd_real = os.path.realpath(cwd)
        wt_real = os.path.realpath(wt_path)
        if cwd_real == wt_real or cwd_real.startswith(wt_real + os.sep):
            os.chdir(main_repo)
            _log(f"🔄 Moved to main repo: {main_repo}")

        # Unlock worktree first (ignore errors - may not be locked)
        subprocess.run(
            ["git", "worktree", "unlock", wt_path],
            capture_output=True,
            timeout=10,
        )

        # Try to remove worktree
        result = subprocess.run(
            ["git", "worktree", "remove", wt_path],
            capture_output=True,
            text=True,
            timeout=30,
        )

        wt_name = os.path.basename(wt_path)

        if result.returncode == 0:
            _log(f"✅ Worktree {wt_path} を削除しました")
            _cleanup_branch(wt_name)
            return True

        # Check if there are uncommitted changes before force removal
        # Force removal is only safe after successful merge (all changes committed)
        error_msg = result.stderr.strip()
        _log(f"⚠️ 通常削除失敗、強制削除を試行: {error_msg}")

        result = subprocess.run(
            ["git", "worktree", "remove", "-f", wt_path],
            capture_output=True,
            text=True,
            timeout=30,
        )

        if result.returncode == 0:
            _log(f"✅ Worktree {wt_path} を強制削除しました")
            _cleanup_branch(wt_name)
            return True

        _log(f"❌ Worktree削除失敗: {result.stderr.strip()}")
        return False

    except subprocess.TimeoutExpired:
        _log("❌ Worktree削除がタイムアウトしました")
        return False
    except Exception as e:  # noqa: BLE001 - Catch all for cleanup errors
        _log(f"❌ Worktree削除エラー: {e}")
        return False

File: examples/dekita/.claude/scripts/ci_monitor_ts/worktree.ts
/**
 * Worktree management for ci-monitor.
 *
 * Why:
 *   Handle git worktree detection and cleanup operations.
 *   Supports cleanup after successful PR merge.
 *
 * What:
 *   - getWorktreeInfo(): Get worktree information if in a worktree
 *   - cleanupWorktreeAfterMerge(): Cleanup worktree after successful merge
 *
 * Remarks:
 *   - Migrated from Python ci_monitor/worktree.py (Issue #3261)
 *   - Uses asyncSpawn for git command execution
 *
 * Changelog:
 *   - silenvx/dekita#3261: TypeScript migration from Python
 */

import { basename, resolve, sep } from "node:path";
import { asyncSpawn } from "../../hooks/lib/spawn";

// =============================================================================
// Types
// =============================================================================

/** Optional log function signature */
export type LogFn = (
  message: string,
  jsonMode: boolean,
  data: Record<string, unknown> | null,
) => void;

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Check if branch name contains exact worktree name as a segment.
 *
 * Examples:
 *   - "fix/issue-1366-cleanup" matches "issue-1366" -> true
 *   - "fix/issue-13669" matches "issue-1366" -> false
 *   - "feature-1234" matches "123" -> false
 *
 * @param branch - Branch name to check
 * @param wtName - Worktree name to match
 * @returns True if the branch contains the exact worktree name as a segment
 */
function isExactWorktreeMatch(branch: string, wtName: string): boolean {
  // Check for exact segment match (surrounded by non-alphanumeric or at boundaries)
  // Escape special regex characters in wtName
  const escaped = wtName.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  // Match only if wtName is a complete segment (not part of a larger number/word)
  const pattern = new RegExp(`(^|[^a-zA-Z0-9])${escaped}([^a-zA-Z0-9]|$)`);
  return pattern.test(branch);
}

// =============================================================================
// Worktree Functions
// =============================================================================

/**
 * Get worktree information if current directory is inside a worktree.
 *
 * @returns Tuple of [mainRepoPath, worktreePath] if in worktree, [null, null] otherwise
 */
export async function getWorktreeInfo(): Promise<[string | null, string | null]> {
  try {
    // Get current working directory
    const cwd = process.cwd();

    // Get git worktree list
    const result = await asyncSpawn("git", ["worktree", "list", "--porcelain"], {
      timeout: 10000,
    });

    if (!result.success || !result.stdout) {
      return [null, null];
    }

    // Parse worktree list
    // Format: worktree /path\nHEAD abc123\nbranch refs/heads/xxx\n\n
    const worktrees: string[] = [];
    let mainRepo: string | null = null;

    for (const line of result.stdout.trim().split("\n")) {
      if (line.startsWith("worktree ")) {
        const path = line.slice(9); // Remove "worktree " prefix
        if (mainRepo === null) {
          mainRepo = path; // First entry is main repo
        } else {
          worktrees.push(path);
        }
      }
    }

    if (mainRepo === null) {
      return [null, null];
    }

    // Check if cwd is inside a worktree (not main repo)
    // Use resolve to normalize paths (Bun.realPathSync is not available in all contexts)
    const cwdResolved = resolve(cwd);
    for (const wtPath of worktrees) {
      const wtResolved = resolve(wtPath);
      if (cwdResolved === wtResolved || cwdResolved.startsWith(wtResolved + sep)) {
        return [mainRepo, wtPath];
      }
    }

    return [null, null];
  } catch {
    return [null, null];
  }
}

/**
 * Cleanup worktree after successful merge.
 *
 * Issue #1366: Prevent worktree from remaining after cleanup phase is skipped.
 *
 * Note: Issue #1478 - This function is no longer called from --merge command
 * because subprocess deletion makes parent process's cwd invalid. However, it
 * is kept for potential use by hooks (e.g., worktree-auto-cleanup.py) that run
 * in the correct process context.
 *
 * @param wtPath - Path to the worktree to remove
 * @param mainRepo - Path to the main repository
 * @param jsonMode - If true, log() outputs to stderr instead of stdout
 * @param logFn - Optional logging function
 * @returns True if cleanup succeeded, false otherwise
 */
export async function cleanupWorktreeAfterMerge(
  wtPath: string,
  mainRepo: string,
  jsonMode = false,
  logFn?: LogFn,
): Promise<boolean> {
  const log = (message: string): void => {
    if (logFn) {
      logFn(message, jsonMode, null);
    } else if (jsonMode) {
      // In JSON mode, use stderr for non-JSON messages to avoid polluting stdout
      console.error(message);
    } else {
      console.log(message);
    }
  };

  const cleanupBranch = async (wtName: string): Promise<void> => {
    try {
      // Get branches with exact issue number match (e.g., issue-1366)
      const branchResult = await asyncSpawn("git", ["branch", "--list"], {
        timeout: 10000,
      });

      if (branchResult.success && branchResult.stdout?.trim()) {
        for (const line of branchResult.stdout.trim().split("\n")) {
          const branch = line.trim().replace(/^\* /, "");
          // Exact match: branch must contain the full wtName as a segment
          if (branch && isExactWorktreeMatch(branch, wtName)) {
            // -d is safe: only deletes fully merged branches
            const delResult = await asyncSpawn("git", ["branch", "-d", branch], {
              timeout: 10000,
            });
            if (delResult.success) {
              log(`🗑️ ブランチ ${branch} を削除しました`);
            }
          }
        }
      }
    } catch (error) {
      // Branch cleanup is best-effort - log error but don't fail
      log(`⚠️ ブランチ削除中にエラー: ${error}`);
    }
  };

  try {
    // If cwd is inside worktree, move to main repo first
    const cwd = process.cwd();
    const cwdResolved = resolve(cwd);
    const wtResolved = resolve(wtPath);

    if (cwdResolved === wtResolved || cwdResolved.startsWith(wtResolved + sep)) {
      process.chdir(mainRepo);
      log(`🔄 Moved to main repo: ${mainRepo}`);
    }

    // Unlock worktree first (ignore errors - may not be locked)
    await asyncSpawn("git", ["worktree", "unlock", wtPath], { timeout: 10000 });

    // Try to remove worktree
    const result = await asyncSpawn("git", ["worktree", "remove", wtPath], {
      timeout: 30000,
    });

    const wtName = basename(wtPath);

    if (result.success) {
      log(`✅ Worktree ${wtPath} を削除しました`);
      await cleanupBranch(wtName);
      return true;
    }

    // Check if there are uncommitted changes before force removal
    // Force removal is only safe after successful merge (all changes committed)
    const errorMsg = result.stderr?.trim() || "";
    log(`⚠️ 通常削除失敗、強制削除を試行: ${errorMsg}`);

    const forceResult = await asyncSpawn("git", ["worktree", "remove", "-f", wtPath], {
      timeout: 30000,
    });

    if (forceResult.success) {
      log(`✅ Worktree ${wtPath} を強制削除しました`);
      await cleanupBranch(wtName);
      return true;
    }

    log(`❌ Worktree削除失敗: ${forceResult.stderr?.trim()}`);
    return false;
  } catch (error) {
    if (error instanceof Error && error.message.includes("timeout")) {
      log("❌ Worktree削除がタイムアウトしました");
    } else {
      log(`❌ Worktree削除エラー: ${error}`);
    }
    return false;
  }
}
